torchao/quantization/quantize_/common/dynamic_activation_quantization_wrapper.py

```diff
@@ -0,0 +1,336 @@
+
+    def __tensor_flatten__(self):
+        return ["original_weight_tensor"], [self.act_tensor_kwargs]
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        original_weight_tensor = tensor_data_dict["original_weight_tensor"]
+        (act_tensor_kwargs,) = tensor_attributes
+        return cls(original_weight_tensor, act_tensor_kwargs)
+
+    @classmethod
+    def from_float(
+        cls,
+        weight: torch.Tensor,
+        act_tensor_kwargs: AOBaseTensorKwargs,
+    ):
+        return cls(weight, act_tensor_kwargs)
+
+    def _apply_fn_to_data(self, fn):
+        return self.__class__(
+            fn(self.original_weight_tensor),
+            self.act_tensor_kwargs,
+        )
+
+    def to(self, *args, **kwargs):
+        kwargs = self._get_to_kwargs(*args, **kwargs)
+        return self.__class__(
+            self.original_weight_tensor.to(**kwargs),
+            self.act_tensor_kwargs,
+        )
+
+
+def _same_metadata(
+    self: DynamicActivationQuantizationWrapper,
+    src: DynamicActivationQuantizationWrapper,
+):
+    return (
+        isinstance(self, DynamicActivationQuantizationWrapper)
+        and isinstance(src, DynamicActivationQuantizationWrapper)
+        and self.shape == src.shape
+        and self.act_tensor_kwargs == src.act_tensor_kwargs
+    )
+
+
+implements = DynamicActivationQuantizationWrapper.implements
+
+
+@implements([torch.nn.functional.linear, aten.linear.default])
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    if isinstance(weight_tensor, DynamicActivationQuantizationWrapper):
+        if input_tensor.numel() == 0:
+            return input_tensor
+        original_weight_tensor = weight_tensor.original_weight_tensor
+        kwargs = weight_tensor.act_tensor_kwargs
+        quantized_input = quantize_tensor(input_tensor, kwargs)
+        return torch.nn.functional.linear(quantized_input, original_weight_tensor, bias)
+
+    raise NotImplementedError(
+        "DynamicActivationQuantizationWrapper: No specialized dispatch found for linear op"
+    )
+
+
+@implements(torch.bmm)
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor = (
+        args[0],
+        args[1],
+    )
+    if isinstance(weight_tensor, DynamicActivationQuantizationWrapper):
+        original_weight_tensor = weight_tensor.original_weight_tensor
+        kwargs = weight_tensor.act_tensor_kwargs
+        quantized_input = quantize_tensor(input_tensor, kwargs)
+        return torch.bmm(quantized_input, original_weight_tensor)
+
+    raise NotImplementedError(
+        "DynamicActivationQuantizationWrapper: No specialized dispatch found for bmm op"
+    )
+
+
+@implements([aten.mm.default, aten.addmm.default])
+def _(func, types, args, kwargs):
```

Comment:
can we make this just quantize the activation and make sure all the other logic (transposing weight, checking dims, etc) is done on the weight tensor override?

Reply:
even further, how about combining the logic for all math ops, such as this:

```python
@implements([mm, addmm, bmm, linear, ...])
def _(...):
    input = args[0]
    act_kwargs = ...
    inner_weight = ...
    quantized_input = _quantize_with_kwargs(input, inner_weight)
    output = func([quantized_input, inner_weight, ...], ...)
```

Reply:
from jerry: sounds good